Revamp Windows docs (#1820)
ansible/ansible-documentation: docs/docsite/rst/os_guide/intro_windows.rst

.. _working_with_windows:

Managing Windows hosts with Ansible
===================================

Managing Windows hosts is different from managing POSIX hosts. If you have managed nodes running Windows, review these topics.

.. contents::
   :local:

Bootstrapping Windows
---------------------

Windows nodes must be running Windows Server 2016 or Windows 10 or newer. As these versions of Windows ship with PowerShell 5.1 by default there are no additional requirements to bootstrap a Windows node.

> Support for each Windows version is tied to the extended support lifecycle of each operating system which is typically 10 years from the date of release. Ansible is tested against the server variants of Windows but should still be compatible with the desktop variants like Windows 10 and 11.

Reviewer: ```suggestion
Support for each Windows version is tied to the extended support lifecycle of each operating system, which is typically 10 years from the date of release. Ansible is tested against the server variants of Windows but should still be compatible with the desktop variants like Windows 10 and 11.
```

There are dedicated Windows modules that are written in PowerShell and are meant to be run on Windows hosts. A list of these modules can be in the :ref:`plugins_in_ansible.windows`, :ref:`plugins_in_community.windows`, :ref:`plugins_in_microsoft.ad`, :ref:`plugins_in_chocolatey.chocolatey`, and other collections.

In addition, the following Ansible Core modules/action-plugins work with Windows:

* add_host
* assert
* async_status
* debug
* fail
* fetch
* group_by
* include
* include_role
* include_vars
* meta
* pause
* raw
* script
* set_fact
* set_stats
* setup
* slurp
* template (also: win_template)
* wait_for_connection

.. _windows_control_node:

Using Windows as the control node
---------------------------------

Ansible cannot run on Windows as the control node due to API limitations on the platform. However, you can run Ansible on Windows using the Windows Subsystem for Linux (``WSL``) or in a container.

.. note::
   The Windows Subsystem for Linux is not supported by Ansible and should not be used for production systems.

Windows facts
-------------

> Ansible gathers facts from Windows in a similar manner to other POSIX hosts but due with some differences. Some facts may be in a different format to be backwards compatible or may not be available at all.

Reviewer: ```suggestion
Ansible gathers facts from Windows in a similar manner to other POSIX hosts but with some differences. Some facts may be in a different format for backwards compatibility or may not be available at all.
```

To see the facts that Ansible gathers from Windows hosts, run the ``setup`` module.

.. code-block:: bash

   ansible windows -m setup

> Common Windows Problems

Reviewer: ```suggestion
Common Windows problems
```

-----------------------

Command works locally but not under Ansible
"""""""""""""""""""""""""""""""""""""""""""

Ansible executes commands through a network logon which can change how Windows authorizes actions. This can cause commands that work locally to fail under Ansible. Some examples of these failures are:

* the process cannot delegate the user's credentials to a network resource, causing ``Access is  Denied`` or ``Resource Unavailable`` errors
* applications that require an interactive session will not work
* some Windows APIs are restricted when running through a network logon
* some tasks require access to the ``DPAPI`` secrets store which is typically not available on a network logon

> The common way is to use :ref:`become` to run a command with explicit credentials. Using ``become`` on Windows will change the network logon to an interactive one and if explicit credentials are provided to the become identity, the command will be able to access network resources and unlock the ``DPAPI`` store.

Reviewer: ```suggestion
The common way is to use :ref:`become` to run a command with explicit credentials. Using ``become`` on Windows will change the network logon to an interactive one and, if explicit credentials are provided to the become identity, the command will be able to access network resources and unlock the ``DPAPI`` store.
```

Another option is to use an authentication option on the connection plugin that allows for credential delegation. For SSH this can be done with an explicit username and password or through a Kerberos/GSSAPI logon with delegation enabled. For WinRM based connections, the CredSSP or Kerberos with delegation can be used. See the connection specific documentation for more information.

Credentials are rejected
""""""""""""""""""""""""

There are a few reasons why credentials might be rejected when connecting to the Windows host. Some common reasons are:

* the username or password is incorrect
* the user account is locked out, disabled, not allowed to log onto that server
* the user account is not allowed to log on through the network
* the user account is not a member of the local Administrators group
* the user account is a local user and the ``LocalAccountTokenFilterPolicy`` is not set

> To verify whether the credentials are correct or the user is allowed to log onto the host you can run the below PowerShell command on the Windows host to see the last failed logon attempt. This will output the even details including the ``Status`` and ``Sub Status`` error code indicating why the logon failed.

Reviewer: ```suggestion
To verify whether the credentials are correct or the user is allowed to log onto the host you can run the below PowerShell command on the Windows host to see the last failed logon attempt. This will output event details including the ``Status`` and ``Sub Status`` error code indicating why the logon failed.
```

.. code-block:: powershell

    Get-WinEvent -FilterHashtable @{LogName = 'Security'; Id = 4625} |
        Select-Object -First 1 -ExpandProperty Message

> While not all connection plugins require the connection user to be a member of the local Administrators group, this is typically the default configuration. If the user user is not a member of the local Administrators group or is a local user without ``LocalAccountTokenFilterPolicy`` set, the authentication will fail.

Reviewer: ```suggestion
While not all connection plugins require the connection user to be a member of the local Administrators group, this is typically the default configuration. If the user is not a member of the local Administrators group or is a local user without ``LocalAccountTokenFilterPolicy`` set, the authentication will fail.
```

.. _working_with_windows:

Managing Windows hosts with Ansible
===================================

Managing Windows hosts is different from managing POSIX hosts. If you have managed nodes running Windows, review these topics.

.. contents::
   :local:

This is an index of all the topics covered in this guide.

.. toctree::
   :maxdepth: 1

>    windows_dsc

Reviewer: I noticed there are some 404s on this page: https://ansible--1820.org.readthedocs.build/projects/ansible-core/en/1820/os_guide/windows_dsc.html#ciminstance-type

While it's not part of this PR, it might be good to fix the broken links while we're in here.

References to class `MSFT_xWebAuthenticationInformation` and class `MSFT_xWebBindingInformation`

- Current link -> https://github.com/dsccommunity/xWebAdministration/blob/master/source/DSCResources/MSFT_xWebSite/MSFT_xWebSite.schema.mof
- Updated link -> https://github.com/dsccommunity/WebAdministrationDsc/blob/main/source/DSCResources/DSC_WebSite/DSC_WebSite.schema.mof

Author: Have updated the links and references to the DSC classes to reference the new type.
